In [1]:
import numpy as np
import pandas as pd

In [2]:
df = pd.DataFrame({'From_To': ['LoNDon_paris', 'MAdrid_miLAN','londON_StockhOlm','Budapest_PaRis', 'Brussels_londOn'],
'FlightNumber': [10045, np.nan, 10065, np.nan, 10085],
'RecentDelays': [[23, 47], [], [24, 43, 87], [13], [67, 32]],
'Airline': ['KLM(!)', '<Air France> (12)', '(British Airways. )','12. Air France', '"Swiss Air"']})

In [3]:
df

From_To  FlightNumber  RecentDelays              Airline
0      LoNDon_paris       10045.0      [23, 47]               KLM(!)
1      MAdrid_miLAN           NaN            []    <Air France> (12)
2  londON_StockhOlm       10065.0  [24, 43, 87]  (British Airways. )
3    Budapest_PaRis           NaN          [13]       12. Air France
4   Brussels_londOn       10085.0      [67, 32]          "Swiss Air"

In [4]:
df.dtypes

From_To          object
FlightNumber    float64
RecentDelays     object
Airline          object
dtype: object

In [5]:
df.describe()

FlightNumber
count           3.0
mean        10065.0
std            20.0
min         10045.0
25%         10055.0
50%         10065.0
75%         10075.0
max         10085.0

1. Some values in the the FlightNumber column are missing. These numbers are
meant to increase by 10 with each row so 10055 and 10075 need to be put in
place. Fill in these missing numbers and make the column an integer column
(instead of a float column).

In [6]:
df['FlightNumber'].interpolate(method='linear', limit_direction='forward',inplace =True)

In [45]:
down_quantiles = df.quantile([0.25,0.75])
down_quantiles

FlightNumber
0.25       10055.0
0.75       10075.0

In [7]:
df

From_To  FlightNumber  RecentDelays              Airline
0      LoNDon_paris       10045.0      [23, 47]               KLM(!)
1      MAdrid_miLAN       10055.0            []    <Air France> (12)
2  londON_StockhOlm       10065.0  [24, 43, 87]  (British Airways. )
3    Budapest_PaRis       10075.0          [13]       12. Air France
4   Brussels_londOn       10085.0      [67, 32]          "Swiss Air"

In [7]:
df

From_To  FlightNumber  RecentDelays              Airline
0      LoNDon_paris       10045.0      [23, 47]               KLM(!)
1      MAdrid_miLAN       10055.0            []    <Air France> (12)
2  londON_StockhOlm       10065.0  [24, 43, 87]  (British Airways. )
3    Budapest_PaRis       10075.0          [13]       12. Air France
4   Brussels_londOn       10085.0      [67, 32]          "Swiss Air"

In [0]:
df['FlightNumber'] = df['FlightNumber'].astype(int)

In [110]:
df

From_To  FlightNumber  RecentDelays              Airline
0      LoNDon_paris         10045      [23, 47]               KLM(!)
1      MAdrid_miLAN         10055            []    <Air France> (12)
2  londON_StockhOlm         10065  [24, 43, 87]  (British Airways. )
3    Budapest_PaRis         10075          [13]       12. Air France
4   Brussels_londOn         10085      [67, 32]          "Swiss Air"

2. The From_To column would be better as two separate columns! Split each
string on the underscore delimiter _ to give a new temporary DataFrame with
the correct values. Assign the correct column names to this temporary
DataFrame.

In [111]:
data = []
df1 = pd.DataFrame(data, columns = ['City1', 'City2'])
df1

Empty DataFrame
Columns: [City1, City2]
Index: []

In [112]:

df1[['City1','City2']] = df.From_To.str.split("_",expand=True)
df1

City1      City2
0    LoNDon      paris
1    MAdrid      miLAN
2    londON  StockhOlm
3  Budapest      PaRis
4  Brussels     londOn

3. Notice how the capitalisation of the city names is all mixed up in this
temporary DataFrame. Standardise the strings so that only the first letter is
uppercase (e.g. "londON" should become "London".)

In [113]:
df1

City1      City2
0    LoNDon      paris
1    MAdrid      miLAN
2    londON  StockhOlm
3  Budapest      PaRis
4  Brussels     londOn

In [0]:
df1['City1'] = df1['City1'].str.capitalize()
df1['City2'] = df1['City2'].str.capitalize()

In [115]:
df1

City1      City2
0    London      Paris
1    Madrid      Milan
2    London  Stockholm
3  Budapest      Paris
4  Brussels     London

4. Delete the From_To column from df and attach the temporary DataFrame
from the previous questions.

In [0]:
df.drop('From_To',axis =1, inplace=True)

In [0]:
Final_df = pd.concat([df, df1], axis=1, sort=False)   

In [118]:
Final_df

FlightNumber  RecentDelays              Airline     City1      City2
0         10045      [23, 47]               KLM(!)    London      Paris
1         10055            []    <Air France> (12)    Madrid      Milan
2         10065  [24, 43, 87]  (British Airways. )    London  Stockholm
3         10075          [13]       12. Air France  Budapest      Paris
4         10085      [67, 32]          "Swiss Air"  Brussels     London

5. In the RecentDelays column, the values have been entered into the
DataFrame as a list. We would like each first value in its own column, each
second value in its own column, and so on. If there isn't an Nth value, the value
should be NaN.
Expand the Series of lists into a DataFrame named delays, rename the columns
delay_1, delay_2, etc. and replace the unwanted RecentDelays column in df
with delays.

In [119]:
df['RecentDelays']

0        [23, 47]
1              []
2    [24, 43, 87]
3            [13]
4        [67, 32]
Name: RecentDelays, dtype: object

In [0]:
df[['Delay_1','Delay_2','Delay_3']] = pd.DataFrame(df.RecentDelays.tolist(), index= df.index)

In [121]:
df

FlightNumber  RecentDelays              Airline  Delay_1  Delay_2  Delay_3
0         10045      [23, 47]               KLM(!)     23.0     47.0      NaN
1         10055            []    <Air France> (12)      NaN      NaN      NaN
2         10065  [24, 43, 87]  (British Airways. )     24.0     43.0     87.0
3         10075          [13]       12. Air France     13.0      NaN      NaN
4         10085      [67, 32]          "Swiss Air"     67.0     32.0      NaN

In [122]:
df.drop('RecentDelays',axis=1,inplace=True)
df

FlightNumber              Airline  Delay_1  Delay_2  Delay_3
0         10045               KLM(!)     23.0     47.0      NaN
1         10055    <Air France> (12)      NaN      NaN      NaN
2         10065  (British Airways. )     24.0     43.0     87.0
3         10075       12. Air France     13.0      NaN      NaN
4         10085          "Swiss Air"     67.0     32.0      NaN